Path update

docwilmot opened the pull request:
Fixes https://github.com/backdrop/backdrop-issues/issues/2077

Allows as per title
Also allows overwriting existing aliases

docwilmot commented:
Last try to get sandbox working.

klonos commented on core/modules/path/path.admin.inc:
```diff
@@ -682,13 +708,54 @@ function path_bulk_update_form() {
+  }
+  $header = array(
+    'alias' => t('Alias'),
+    'delete' => t('Delete existing'),
```
As discussed in the issue, better to be `'Update existing'`

jenlampton commented on core/modules/path/css/path.css:
```diff
@@ -0,0 +1,10 @@
+.path-type {
+  margin-left: 40px;
```
Do we need to add some RTL styles for this too?

jenlampton commented on core/modules/path/path.module:
```diff
@@ -92,25 +92,31 @@ function path_menu() {
   );
   $items['admin/config/search/path/bulk-update'] = array(
-    'title' => 'Bulk generate',
+    'title' => 'Bulk alias actions',
```
How about `Update URL aliases` here? match Delete?

jenlampton commented on core/modules/path/path.path.inc:
```diff
@@ -18,6 +18,7 @@ function node_path_info() {
     'pattern description' => t('Default URL alias pattern (applies to all content types with blank URL alias patterns below)'),
     'pattern default' => 'content/[node:title]',
     'batch update callback' => 'node_path_bulk_update_batch_process',
+    'type delete callback' => 'node_path_type_delete_callback',
```
We should probably batch process these too? Maybe...

```
'batch delete callback' => 'node_path_bulk_delete_batch_process',
```

docwilmot replied:
Wondered about that. It wasn't batched in the pre-PR code either.
We batch the update callback because its a query, followed by a node load on each query result, followed by      path_save_automatic_entity_alias() on each node. Thats heavy. But this is just a single DB delete query, so maybe.

jenlampton commented on core/modules/path/path.path.inc:
```diff
@@ -47,6 +48,26 @@ function node_path_info() {
+  }
+
+  $query = db_delete('url_alias');
+  $query->condition('source', $sources, 'IN');
```
Hm, this looks pretty simple, Maybe we don't need a batch. @quicksketch ?

quicksketch replied:
This could cause a problem with anything more than a few thousand items. I was going to recommend a subquery here, but it looks like we need to tack on `node/` onto each path before we do the `IN` condition. So yes... looks like a batch is necessary.

jenlampton commented on core/modules/path/path.path.inc:
```diff
@@ -105,6 +146,7 @@ function taxonomy_path_info() {
     'pattern description' => t('Default URL alias pattern (applies to all vocabularies with blank URL alias patterns below)'),
     'pattern default' => '[term:vocabulary]/[term:name]',
     'batch update callback' => 'taxonomy_path_bulk_update_batch_process',
+    'type delete callback' => 'taxonomy_path_type_delete_callback',
```
to batch or not to batch, that is the question...

docwilmot commented:
Tests are all green now, but currently woefully inadequate. It seems we never even had any tests for the UI bulk delete facility!! Would file a new issue, but I think i can just create new tests to fix that with this PR.

docwilmot commented:
Tests upgraded and comprehensive. now awaiting final review

klonos commented:
The checking/unchecking of checkboxes is now more consistent and there is a (un)select-all checkbox for the "Update existing" column that also works great with (un)selecting. Ticking and unticking boxes that apply to groups also works as expected now 👍 

The only difference I see with my previous round of testing is that all the checkboxes in the "Update existing" column are enabled even though their respective "Alias" column checkboxes are not ticked. This causes for example situations where one can select something to be updated even though its respective checkbox to execute the bulk update is not selected in the first place.

klonos commented:
...also, from what I saw, tests are not testing the "Update existing" functionality. Could be wrong though.

docwilmot commented:
Right and right.

I gave up on the disabling checkboxes. The PHP code prevents you from accomplishing anything if you click an update checkbox but leave the alias unchecked anyway. But the JQuery was driving me insane. I leave out the disable JQuery stuff until I get some inspiration or somebody who can do it better (hint: @serundeputy @quicksketch) takes up the challenge. But a validation check may be a good idea there anyway; I'll work on that.

And embarrassingly I forgot to test the update existing. Will fix that too. 

Thanks

quicksketch commented on core/modules/path/path.path.inc:
```diff
@@ -121,6 +163,25 @@ function taxonomy_path_info() {
+  }
+
+  $query = db_delete('url_alias');
+  $query->condition('source', $sources, 'IN');
```
Same as above, this query could be positively huge. ~~We'll need to batch.~~

See https://github.com/backdrop/backdrop-issues/issues/2077#issuecomment-245194253.

docwilmot commented:
OK, final round (hopefully).
- fixed the bogus counting of aliases ('twas truly bogus)
- updated tests to test reset functionality
- added validation functions
- updated API docs.

jenlampton commented:
Batching?
- -
  Please excuse brevity and typos. Message typed with my thumbs from an Open
  Source OS.

On Sep 8, 2016 6:40 PM, "docwilmot" notifications@github.com wrote:

> OK, final round (hopefully).
> - fixed the bogus counting of aliases ('twas truly bogus)
> - updated tests to test reset functionality
> - added validation functions
> - updated API docs.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/backdrop/backdrop/pull/1489#issuecomment-245794245,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAYSR6KikUUcdvmzVQYk6YAVLLHZeQEwks5qoLkegaJpZM4JoEVT
> .

docwilmot commented:
[@quicksketch says it can wait](https://github.com/backdrop/backdrop-issues/issues/2077#issuecomment-245194253) and @jenlampton agreed with him. :smile:

jenlampton commented:
Well I defer to them! :)

quicksketch commented:
This needs a little work from the JS-side. I'm going to try cleaning it up a bit and refiling.

docwilmot commented:
> This needs a little work from the JS-side

Believe me, I know!

quicksketch commented:
New PR at https://github.com/backdrop/backdrop/pull/1533 replaces this one with a few changes:
- New `path_generate.js` moved to `path.admin.js`.
- New `path.css` moved to `path.admin.css`.
- Fixed jQuery variables in the global namespace and added $.once() checking to prevent double-binding of events.
- Added RTL styles and tidied up the CSS to be page-specific in the event `path.admin.css` gets used elsewhere at some point.
